### PR TITLE
ECOM-3464: Added the command to publish the courses to LMS

### PR DESCRIPTION
**Pull request by aamir-khan**
Added the command to take the course ids as input in text files and publish these course to LMS

Related Ticket: [ECOM-3464](https://openedx.atlassian.net/browse/ECOM-3464)

@adampalay Can you please review this PR.

### CODE REVIEW

**Comment by aamir-khan** [comment]
@edx/ecommerce @rlucioni @clintonb @bderusha @peter-fogg Can you please review this PR.

File: ecommerce/courses/management/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'arbisoft'
```
Comment (adampalay) [inline ecommerce/courses/management/__init__.py]: this line should be removed

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    )
+
+    def handle(self, *args, **options):
+        errors_dict = {}
+        course_ids_file = options['course_ids_file']
+        if not course_ids_file or not os.path.exists(course_ids_file):
+            raise CommandError("Pass the correct absolute path to course ids file as --course_ids_file argument.")
+
+        with open(course_ids_file, 'r') as course_ids:
+            self.stderr.write("Publishing courses to LMS")
```
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: why write to stderr? Why not just print?

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            for course_id in course_ids:
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id.strip())
```
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `course_id` was already stripped in the line above

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+from ecommerce.courses.models import Course
+from ecommerce.courses.tests.factories import CourseFactory
+from ecommerce.tests.testcases import TestCase
+
+
+@ddt.ddt
+class PublishCoursesToLMSTests(TestCase):
+    """Tests the course publish command."""
+
+    tmp_file_path = os.path.join(tempfile.gettempdir(), "tmp-testfile.txt")
```
Comment (adampalay) [inline ecommerce/courses/tests/test_publish_to_lms_command.py:29]: would this tempfile be cleared after tests are run? It isn't clear to me that it would

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        self.course = CourseFactory()
+        self.create_course_ids_file(self.course.id)
+
+    def create_course_ids_file(self, course_id):
+        with open(self.tmp_file_path, 'w') as f:
+            f.write(course_id)
+
+    @ddt.data("", "fake/path")
+    def test_invalid_file_path(self, course_ids_file):
+        """ The command should raise CommandError if invalid path is passed. """
```
Comment (adampalay) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: no need for doc strings in these tests (the function names are descriptive enough)

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                    publishing_error = course.publish_to_lms()
+                    if publishing_error:
+                        errors_dict[course_id] = publishing_error
+                except Course.DoesNotExist:
+                    errors_dict[course_id] = "The course {course_id} does not exists.".format(course_id=course_id)
+        if errors_dict:
+            error_str = [
+                "Course_id={} Error={}".format(course_id, error_msg) for course_id, error_msg in errors_dict.items()
+                ]
```
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: de-indent this bracket

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            error_str = "\n".join(error_str)
+            error_str = "\n".join(["Following Courses failed while publishing through management command:", error_str])
+            self.stderr.write(error_str)
+            logger.error(error_str)
+        else:
+            self.stderr.write("All Courses published successfully.")
```
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: @peter-fogg , @bderusha , strong opinions about whether or not this should be logged or printed?

**Comment by adampalay** [comment]
@aamir-khan , first pass complete

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 2 above)
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Use standard Python logging.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+""" This command publish the courses to LMS."""
+
+from __future__ import with_statement
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Why is this necessary?

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 5 above)
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: I disagree. I prefer tests to be explicit up front about what they are testing.

**Comment by clintonb** [comment]
Why was it decided to use file input rather than re-publish all courses?

**Comment by adampalay** [comment]
@clintonb , to give us flexibility if the audit script didn't give back a clean bill of health

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 4 above)
Comment (aamir-khan) [inline ecommerce/courses/tests/test_publish_to_lms_command.py:29]: Added the tearDownClass method to clear the tempfile on test suit completion.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    def setUp(self):
+        super(PublishCoursesToLMSTests, self).setUp()
+        self.course = CourseFactory()
+        self.create_course_ids_file(self.course.id)
+
+    @classmethod
+    def tearDownClass(cls):
+        if os.path.exists(cls.tmp_file_path):
+            try:
```
Comment (adampalay) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: why a try/except here? If this wasn't working, I think we'd want the test to fail so we'd know about it

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 2 above)
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: @aamir-khan ^^

Comment (cpennington) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Actually, for management commands, using `self.stderr` and `self.stdout` to display messages to the user is the correct usage.

Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: We want this information to go to Splunk, which is setup using the Python logging.

Comment (cpennington) [inline ecommerce/courses/management/commands/publish_to_lms.py]: This message in particular seems like a bad one to go to splunk, since it has no interesting data in it. (Except maybe the timestamp?)

Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Splunk filters to INFO severity or higher. If this message is not important, use `log.debug`.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            self.stderr.write("Publishing courses to LMS")
+            for course_id in course_ids:
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id)
+                    publishing_error = course.publish_to_lms()
+                    if publishing_error:
+                        errors_dict[course_id] = publishing_error
+                except Course.DoesNotExist:
+                    errors_dict[course_id] = "The course {course_id} does not exists.".format(course_id=course_id)
```
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Wording: "does not exist.".

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        if errors_dict:
+            error_str = ["Course_id={} Error={}".format(course_id, error_msg) for course_id, error_msg
+                         in errors_dict.items()]
+            error_str = "\n".join(["Following Courses failed while publishing through management command:"] + error_str)
+            self.stderr.write(error_str)
+            logger.error(error_str)
```
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: I don't see any reason to both log and write to `stderr`.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        else:
+            self.stderr.write("All Courses published successfully.")
+            logger.info("All Courses published successfully.")
+        self.stderr.write("Done.")
+        logger.info("Done.")
```
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Same here.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                os.remove(cls.tmp_file_path)
+            except:  # pylint: disable=bare-except
+                logger.error("Error on deleting the temp file %s", cls.tmp_file_path)
```
Comment (peter-fogg) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: I agree with @adampalay above. Though, if we are going to catch the exception here, we should at least log its message.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+    def create_course_ids_file(self, course_id):
+        """Write the course_id to the temp file."""
+
+        with open(self.tmp_file_path, 'w') as f:
+            f.write(course_id)
```
Comment (peter-fogg) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Might be worth testing the command's file parsing by using multiple course IDs.

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 13 above)
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: `logger.exception()`

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 7 above)
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: For this message, I don't see any reason I'd ever want to find it in Splunk. As such I lean towards printing.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        if not course_ids_file or not os.path.exists(course_ids_file):
+            raise CommandError("Pass the correct absolute path to course ids file as --course_ids_file argument.")
+
+        with open(course_ids_file, 'r') as course_ids:
+            self.stderr.write("Publishing courses to LMS")
+            for course_id in course_ids:
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id)
+                    publishing_error = course.publish_to_lms()
```
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py:47]: On the subject of logging, this would be a good place to note that we're publishing this course through the management command (and not through the normal publication endpoint).

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 11 above)
Comment (aamir-khan) [inline ecommerce/courses/management/commands/publish_to_lms.py]: @peter-fogg My assumption for using both log and stderr is that stderr is for user running the command to see what happened to command he ran and the command progress on terminal instead of searching for logs on splunk. The log is used for splunk logging to check later if anything need to be verified. If it is not correct I can remove one of these logging. @adampalay

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 9 above)
Comment (aamir-khan) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Agreed removing the try/except.

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 11 above)
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: You can add a handler to send log output to the console. See https://docs.python.org/3.1/library/logging.html#configuring-logging.

**Comment by aamir-khan** [comment]
@edx/ecommerce @rlucioni @clintonb @bderusha @peter-fogg @adampalay All comments have been addressed. Can you give this another review please?

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                    if publishing_error:
+                        errors_dict[course_id] = publishing_error
+                except Course.DoesNotExist:
+                    errors_dict[course_id] = "The course {course_id} does not exist.".format(course_id=course_id)
+        if errors_dict:
+            error_str = ["Course_id={} Error={}".format(course_id, error_msg) for course_id, error_msg
+                         in errors_dict.items()]
+            error_str = "\n".join(["Following Courses failed while publishing through management command:"] + error_str)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: The use of `str.join()` here is somewhat confusing. Try this:

``` python
error_msg = "Errors occurred for the following courses while publishing:\n"
for course_id, error in errors_dict.items():
  error_msg += "{course_id}\t{error}\n".format(course_id=course_id, error=error)
```

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
+"""Contains the tests for publish to lms command."""
+
+import logging
+import os
+import tempfile
+
+import ddt
+from django.core.management import call_command
+from django.core.management import CommandError
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Combine into a single import.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
+""" This command publish the courses to LMS."""
+
+import logging
+from optparse import make_option
+import os
+
+from django.core.management import BaseCommand, CommandError
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Add a newline after this import.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+    @classmethod
+    def tearDownClass(cls):
+        if os.path.exists(cls.tmp_file_path):
+            os.remove(cls.tmp_file_path)
+
+    def create_course_ids_file(self, course_ids):
+        """Write the course_ids list to the temp file."""
+
+        with open(self.tmp_file_path, 'w') as f:
+            # for course_id in course_ids:
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Remove dead code.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    def test_invalid_file_path(self, course_ids_file):
+        """ The command should raise CommandError if invalid path is passed. """
+
+        with self.assertRaises(CommandError):
+            call_command('publish_to_lms', course_ids_file=course_ids_file)
+
+    def test_invalid_course_id(self):
+        """ The course failed due to invalid course id. """
+
+        # Mock the LMS call
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Remove.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+    @ddt.data("", "fake/path")
+    def test_invalid_file_path(self, course_ids_file):
+        """ The command should raise CommandError if invalid path is passed. """
+
+        with self.assertRaises(CommandError):
+            call_command('publish_to_lms', course_ids_file=course_ids_file)
+
+    def test_invalid_course_id(self):
+        """ The course failed due to invalid course id. """
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: This is not a valid docstring for a test case. You should describe what the test validates.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+                        'Error=The course fake_course_id does not exist.'.format(fake_course_id)
+                    )
+                )
+
+    def test_course_publish_successfully(self):
+        """ All courses should be published successfully."""
+
+        second_course = CourseFactory.create()
+        self.create_course_ids_file([self.course.id, second_course.id])
+        # Mock the LMS call
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Remove.

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 20 above)
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Remove. This comment isn't necessary as it is somewhat obvious what the mock is mocking.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+                    (
+                        LOGGER_NAME,
+                        'ERROR',
+                        'Following Courses failed while publishing through management command:\nCourse_id={} '
+                        'Error=The course fake_course_id does not exist.'.format(fake_course_id)
+                    )
+                )
+
+    def test_course_publish_successfully(self):
+        """ All courses should be published successfully."""
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Verify all courses are successfully published.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+                        'All Courses published successfully.'
+                    )
+                )
+        # Check that the mocked function was called twice.
+        self.assertListEqual(mock_publish.call_args_list, [call(self.course), call(second_course)])
+
+    def test_course_publish_failed(self):
+        """ The course failed to be published."""
+
+        # Mock the LMS call
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: Remove.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+                        errors_dict[course_id] = publishing_error
+                except Course.DoesNotExist:
+                    errors_dict[course_id] = "The course {course_id} does not exist.".format(course_id=course_id)
+        if errors_dict:
+            error_str = ["Course_id={} Error={}".format(course_id, error_msg) for course_id, error_msg
+                         in errors_dict.items()]
+            error_str = "\n".join(["Following Courses failed while publishing through management command:"] + error_str)
+            logger.error(error_str)
+        else:
+            logger.debug("All Courses published successfully.")
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `logger.info`

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+        with mock.patch.object(Course, 'publish_to_lms') as mock_publish:
+            error_msg = "The failure message."
+            mock_publish.return_value = error_msg
+            with LogCapture(LOGGER_NAME) as l:
```
Comment (adampalay) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: nit: avoid single-character variable names. Can you do `log_capture` or something more descriptive?

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+    def test_course_publish_successfully(self):
+        """ All courses should be published successfully."""
+
+        second_course = CourseFactory.create()
+        self.create_course_ids_file([self.course.id, second_course.id])
+        # Mock the LMS call
+        with mock.patch.object(Course, 'publish_to_lms', autospec=True) as mock_publish:
+            mock_publish.return_value = None
+            with LogCapture(LOGGER_NAME) as l:
```
Comment (adampalay) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: see below about one-character variables

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 15 above)
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py:47]: @peter-fogg , good point. Should we also log each time a course is published?

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 16 above)
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: On second thought, I don't like the fact that we are waiting until the end of the process to log errors, or anything for that matter. If I am using this, and it takes 5 minutes to run, I see no output for five minutes. The command should log the state of every publication attempt immediately after said attempt. That way, if the first few attempts fail, we can stop the command and fix the issue.

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 26 above)
Comment (clintonb) [inline ecommerce/courses/tests/test_publish_to_lms_command.py]: We normally use `lc`.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+        with open(course_ids_file, 'r') as course_ids:
+            for course_id in course_ids:
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id)
+                    publishing_error = course.publish_to_lms()
+                    if publishing_error:
+                        errors_dict[course_id] = publishing_error
+                except Course.DoesNotExist:
+                    errors_dict[course_id] = "The course {course_id} does not exist.".format(course_id=course_id)
```
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: This string should be unicode (`u"The course ..."`), since course IDs can have non-ASCII characters. Same with all other strings in this file -- using `from __future__ import unicode_literals` might be a good idea.

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 15 above)
Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py:47]: Yeah, that's my point -- I think `logger.info` would be the right logging level.

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 16 above)
Comment (aamir-khan) [inline ecommerce/courses/management/commands/publish_to_lms.py]: @clintonb I agree with your suggestion for logging the state of every publication. I think, while iterating the courses we should log the state of every publication(success/failure) with DEBUG level that will only be displayed on terminal and will not go to splunk. All the failed publications should be logged at the end with ERROR level so that it will be easy to check which courses failed after the command completed.

Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Why is everyone so afraid to log to Splunk. We _want_ to keep track of these logs, not let them die on the respective servers. Please log successes as INFO. Knowing which courses succeeded as just as important as knowing which failed.

File: ecommerce/courses/tests/test_publish_to_lms_command.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+            os.remove(cls.tmp_file_path)
+
+    def create_course_ids_file(self, course_ids):
+        """Write the course_ids list to the temp file."""
+
+        with open(self.tmp_file_path, 'w') as temp_file:
+            temp_file.write("\n".join(course_ids))
+
+    @ddt.data("", "fake/path")
+    def test_invalid_file_path(self, course_ids_file):
```
Comment (awaisdar001) [inline ecommerce/courses/tests/test_publish_to_lms_command.py:48]: Can we also verify file nam with special chars and unicode chars?

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+                total_courses += 1
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id)
+                    publishing_error = course.publish_to_lms()
+                    if publishing_error:
+                        failed += 1
+                        logger.error(u"%s\t%s\n", course_id, publishing_error)
+                    else:
+                        msg = "The course published successfully."
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: This variable is not necessary.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+                        logger.info(u"%s\t%s\n", course_id, msg)
+                except Course.DoesNotExist:
+                    failed += 1
+                    msg = u"The course {} does not exist.".format(course_id)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: This variable is not necessary.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+
+    def handle(self, *args, **options):
+        total_courses = 0
+        failed = 0
+        course_ids_file = options['course_ids_file']
+        if not course_ids_file or not os.path.exists(course_ids_file):
+            raise CommandError("Pass the correct absolute path to course ids file as --course_ids_file argument.")
+
+        with open(course_ids_file, 'r') as course_ids:
+            for course_id in course_ids:
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Log the total count early, so those running the command have an instant sanity check regarding how many courses are being loaded. In general, long-running commands should be somewhat verbose to avoid people thinking the command has somehow stalled or frozen. As @rock345 likes to say, over-communicate rather than under-communicate.

Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Log the total count early, so those running the command have an instant sanity check regarding how many courses are being loaded. In general, long-running commands should be somewhat verbose to avoid people thinking the command has somehow stalled or frozen. As @rock345 likes to say, over-communicate rather than under-communicate.

File: ecommerce/courses/tests/test_publish_to_lms_command.py (hunk 29 above)
Comment (aamir-khan) [inline ecommerce/courses/tests/test_publish_to_lms_command.py:48]: @awaisdar001 I added a test for this too.

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 30 above)
Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: A note on the logging message — can you provide more context? If we're looking this up in splunk, this log is just not going to be useful. Which course published successfully? We should also know that it failed while running a management command, as opposed to failing during a regular publishing event.

**Comment by adampalay** [comment]
@aamir-khan , I suggested improvements to the logging. We want to know more context:
- these logs are being thrown by this management command
- which course is being discussed

If you had to investigate a bug, what is the format of a bug that would be useful to you?

File: ecommerce/courses/management/commands/publish_to_lms.py (hunk 32 above)
Comment (aamir-khan) [inline ecommerce/courses/management/commands/publish_to_lms.py]: @clintonb to get the total count we have to call readlines() function that will load all the data in memory instead of using the generator and may cause the problem if the file size is large.

Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: That shouldn't be more than 2000 strings today. It'll be fine.

**Comment by aamir-khan** [comment]
@adampalay I have updated the logging.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 33]
```diff
@@ -0,0 +1,64 @@
...
+                            course_id, publishing_error)
+                    else:
+                        logger.info(u"The course %s published successfully through management command.\n", course_id)
+                except Course.DoesNotExist:
+                    failed += 1
+                    logger.error(
+                        u"Error in publishing course %s through management command.\tError detail: The course %s does "
+                        u"not exist.\n", course_id, course_id)
+        if failed:
+            logger.error("Management Command: %s course(s) failed out of %s.", failed, total_courses)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: It will be obvious from the logs that this message originated from a management command. The logger name will be `ecommerce.courses.management.commands.publish_to_lms`. There is no need to point out that these messages are coming from a management command anywhere.

Comment (peter-fogg) [inline ecommerce/courses/management/commands/publish_to_lms.py]: I realize this may be due to confusion over my comment above -- to clarify, Clinton's right that we don't need to point out that logging from this file is coming from a management command. I meant that we should be doing some logging here, and not just relying on whatever logging happens in the LMS.

Comment (adampalay) [inline ecommerce/courses/management/commands/publish_to_lms.py]: Ok, sounds good to me

**Comment by adampalay** [comment]
@aamir-khan , sorry to turn this around, but @clintonb pointed out in a previous comment that we'd know what file this code was run in, so it wasn't as important to add information that this was run as a management command to the logs. That being said, still :+1: from me.

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+    def handle(self, *args, **options):
+        failed = 0
+        course_ids_file = options['course_ids_file']
+        if not course_ids_file or not os.path.exists(course_ids_file):
+            raise CommandError("Pass the correct absolute path to course ids file as --course_ids_file argument.")
+
+        with open(course_ids_file, 'r') as file_handler:
+            course_ids = file_handler.readlines()
+            total_courses = len(course_ids)
+            logger.info("Publishing %s course(s) through management command.", total_courses)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: There are much simpler ways to publish a single course.

`Publishing %s courses...`

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+            for index, course_id in enumerate(course_ids, start=1):
+                try:
+                    course_id = course_id.strip()
+                    course = Course.objects.get(id=course_id)
+                    publishing_error = course.publish_to_lms()
+                    if publishing_error:
+                        failed += 1
+                        logger.error(
+                            u"(%s/%s) Error in publishing course %s through management command.\tError detail: %s\n",
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `(%d/%d) Failed to publish %s: %s`

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 36]
```diff
@@ -0,0 +1,64 @@
...
+                            index, total_courses, course_id, publishing_error)
+                    else:
+                        logger.info(u"(%s/%s) The course %s published successfully through management command.\n",
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `(%d/%d) Successfully published %s.`

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+                                    index, total_courses, course_id)
+                except Course.DoesNotExist:
+                    failed += 1
+                    logger.error(
+                        u"(%s/%s) Error in publishing course %s through management command.\tError detail: "
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `(%d/%d) Failed to publish %s: Course does not exist.`

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 38]
```diff
@@ -0,0 +1,64 @@
...
+                        u"The course %s does not exist.\n", index, total_courses, course_id, course_id)
+        if failed:
+            logger.error("%s course(s) failed out of %s.", failed, total_courses)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `Completed publishing courses. %d of %d failed.`

File: ecommerce/courses/management/commands/publish_to_lms.py [hunk 39]
```diff
@@ -0,0 +1,64 @@
...
+        else:
+            logger.info("All %s course(s) published successfully.", total_courses)
```
Comment (clintonb) [inline ecommerce/courses/management/commands/publish_to_lms.py]: `All %d courses successfully published.`

**Comment by clintonb** [comment]
:+1: (once log statements updated)
